File: pegasusX/apps/warehouse-app-ios/WarehouseApp/Views/Dispatch/DispatchOrderList.swift
import SwiftUI

/// Reusable orders segment for the Dispatch screen.
///
/// Renders fleet truck cards, dispatch-mode picker (Smart / Manual),
/// driver picker, undispatched order list with selection checkboxes,
/// smart-suggest preview warnings, and proposed routes.
struct DispatchOrderList: View {
    let preview: DispatchPreview
    let fleetVehicles: [Vehicle]
    @Binding var vehicleReasons: [String: String]
    @Binding var vehicleNotes: [String: String]
    let mutatingFleetVehicleId: String?
    @Binding var dispatchMode: DispatchAssignmentMode
    @Binding var selectedDriverId: String
    @Binding var selectedOrderIds: Set<String>
    let executing: Bool

    // Callbacks
    var onManualDispatch: () -> Void
    var onSmartDispatch: () -> Void
    var onProposeDate: (_ orderId: String) -> Void
    var onReject: (_ orderId: String) -> Void
    var onOrderDoubleTap: (_ orderId: String) -> Void
    var onMarkVehicleUnavailable: (_ vehicle: Vehicle) -> Void
    var onRestoreVehicle: (_ vehicle: Vehicle) -> Void

    private let dispatchTetrisBuffer: Double = 0.95

    var body: some View {
        let selectedDriver = preview.availableDrivers.first(where: { $0.driverId == selectedDriverId })
        let selectedVolume = preview.undispatchedOrders
            .filter { selectedOrderIds.contains($0.orderId) }
            .reduce(0.0) { $0 + $1.volumeVu }
        let effectiveMax: Double = {
            guard let driver = selectedDriver else { return 0 }
            if let free = driver.freeVolumeVu, free > 0 {
                return free * dispatchTetrisBuffer
            }
            return driver.maxVolumeVu * dispatchTetrisBuffer
        }()

        ResponsiveGridContentWrapper {
            // ── Fleet trucks ──
            if !fleetVehicles.isEmpty {
                Section(L10n.format("mobile_warehouse.ui.fleet_trucks_count", "\(fleetVehicles.count)")) {
                    ForEach(fleetVehicles) { vehicle in
                        let reasonBinding = Binding<String>(
                            get: { vehicleReasons[vehicle.vehicleId] ?? vehicle.unavailableReason ?? "MANUAL_HOLD" },
                            set: { vehicleReasons[vehicle.vehicleId] = $0 }
                        )
                        let noteBinding = Binding<String>(
                            get: { vehicleNotes[vehicle.vehicleId] ?? vehicle.unavailableNote ?? "" },
                            set: { vehicleNotes[vehicle.vehicleId] = $0 }
                        )
                        FleetTruckDispatchCard(
                            vehicle: vehicle,
                            selectedReason: reasonBinding,
                            customNote: noteBinding,
                            mutating: mutatingFleetVehicleId == vehicle.vehicleId,
                            onMarkUnavailable: { onMarkVehicleUnavailable(vehicle) },
                            onRestore: { onRestoreVehicle(vehicle) }
                        )
                    }
                }
            }

            // ── Empty state ──
            if preview.undispatchedOrders.isEmpty {
                Section {
                    ContentUnavailableView("All Dispatched", systemImage: "checkmark.circle", description: Text("mobile_warehouse.ui.no_pending_orders"))
                }
            } else {
                // ── Dispatch mode selector ──
                Section {
                    Picker("Dispatch mode", selection: $dispatchMode) {
                        ForEach(DispatchAssignmentMode.allCases) { mode in
                            Text(mode.label).tag(mode)
                        }
                    }
                    .pickerStyle(.segmented)
                    if dispatchMode == .manual {
                    Picker("Truck / driver", selection: $selectedDriverId) {
                        Text("mobile_warehouse.ui.select_truck_driver").tag("")
                        ForEach(preview.availableDrivers) { driver in
                            Text(driverPickerLabel(driver)).tag(driver.driverId)
                        }
                    }
                    if selectedDriver != nil {
                        Text(L10n.format("mobile_warehouse.ui.loaded_n_1f_n_1f_2_vu_effective", String(format: "%.1f", selectedVolume), String(format: "%.1f", effectiveMax)))
                            .font(.subheadline)
                            .foregroundStyle(.secondary)
                    }
                    Button {
                        onManualDispatch()
                    } label: {
                        Text(executing ? "Dispatching…" : "Manual (\(selectedOrderIds.count))")
                            .frame(maxWidth: .infinity)
                    }
                    .buttonStyle(.borderedProminent)
                    .disabled(executing || selectedDriverId.isEmpty || selectedOrderIds.isEmpty)
                    } else {
                    Text("mobile_warehouse.ui.trucks_are_assigned_automatically_across_the_fleet")
                        .font(.subheadline)
                        .foregroundStyle(.secondary)
                    Button {
                        onSmartDispatch()
                    } label: {
                        Text("mobile_warehouse.ui.smart_dispatch")
                            .frame(maxWidth: .infinity)
                    }
                    .buttonStyle(.borderedProminent)
                    .disabled(executing || preview.undispatchedOrders.isEmpty || preview.availableDrivers.isEmpty)
                    }
                    if preview.fleetEffectiveCapacityVu > 0 {
                        Text(L10n.format("mobile_warehouse.ui.fleet_n_1f_vu_effective", String(format: "%.1f", preview.fleetEffectiveCapacityVu)))
                            .font(.subheadline)
                            .foregroundStyle(.secondary)
                    }
                }

                // ── Orders list ──
                Section("Orders") {
                    ForEach(preview.undispatchedOrders) { order in
                        HStack(alignment: .center, spacing: LabTheme.spacingSM) {
                            Button {
                                if selectedOrderIds.contains(order.orderId) {
                                    selectedOrderIds.remove(order.orderId)
                                } else {
                                    selectedOrderIds.insert(order.orderId)
                                }
                            } label: {
                                Image(systemName: selectedOrderIds.contains(order.orderId) ? "checkmark.circle.fill" : "circle")
                                    .foregroundStyle(selectedOrderIds.contains(order.orderId) ? Color.accentColor : .secondary)
                            }
                            .buttonStyle(.plain)

                            VStack(alignment: .leading, spacing: LabTheme.spacingXS) {
                                Text(order.retailerName.isEmpty ? String(order.orderId.prefix(8)) : order.retailerName)
                                    .font(.headline)
                                Text(L10n.format("mobile_warehouse.ui.formatted_uzs_n_1f_vu", "\(order.totalUzs.formatted())", String(format: "%.1f", order.volumeVu ?? 0)))
                                    .font(.subheadline)
                                    .foregroundStyle(.secondary)
                                HStack(spacing: LabTheme.spacingSM) {
                                    Button("mobile_warehouse.ui.propose_date") {
                                        onProposeDate(order.orderId)
                                    }
                                    .buttonStyle(.bordered)
                                    .controlSize(.small)
                                    Button("common.action.cancel", role: .destructive) {
                                        onReject(order.orderId)
                                    }
                                    .buttonStyle(.bordered)
                                    .controlSize(.small)
                                }
                            }
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .contentShape(Rectangle())
                            .onTapGesture(count: 2) {
                                onOrderDoubleTap(order.orderId)
                            }
                            .accessibilityHint("Double-tap to open order detail")
                        }
                    }
                }

                // ── Smart suggest preview ──
                if !preview.proposedRoutes.isEmpty || !preview.optimizerWarnings.isEmpty || preview.windowConstrainedCount > 0 {
                    Section("Smart suggest preview") {
                        if preview.windowConstrainedCount > 0 {
                            Text(L10n.format("mobile_warehouse.ui.windowconstrainedcount_order_s_constrained_by_receiving_window", "\(preview.windowConstrainedCount)"))
                                .font(.caption)
                                .foregroundStyle(.orange)
                        }
                        if let source = preview.optimizerSource, !source.isEmpty {
                            Text(L10n.format("mobile_warehouse.ui.source_source_2", "\(source)"))
                                .font(.caption)
                                .foregroundStyle(.secondary)
                        }
                        ForEach(preview.optimizerWarnings, id: \.self) { warning in
                            Text(warning)
                                .font(.caption)
                                .foregroundStyle(.orange)
                        }
                    }
                }

                // ── Proposed routes ──
                if !preview.proposedRoutes.isEmpty {
                    Section(L10n.format("mobile_warehouse.ui.smart_suggest_routes_count", "\(preview.proposedRoutes.count)")) {
                        DispatchPreviewMapView(routes: preview.proposedRoutes)
                            .listRowInsets(EdgeInsets())
                        ForEach(preview.proposedRoutes) { route in
                            VStack(alignment: .leading, spacing: LabTheme.spacingXS) {
                                HStack {
                                    Text(route.driverName ?? route.driverId ?? "Driver")
                                        .font(.headline)
                                    Spacer()
                                    Text(L10n.format("mobile_warehouse.ui.count_stops_n_1f_vu", "\(route.stopCount ?? route.orderIds.count)", String(format: "%.1f", route.volumeVu ?? route.loadedVolume ?? 0)))
                                        .font(.caption)
                                        .foregroundStyle(.secondary)
                                }
                                Text(route.orderIds.joined(separator: " → "))
                                    .font(.caption.monospaced())
                                    .lineLimit(2)
                            }
                        }
                    }
                }
            }
        }
    }

    private func driverPickerLabel(_ driver: AvailableDriver) -> String {
        let vehicle = driver.vehicleLabel.isEmpty ? driver.truckStatus : driver.vehicleLabel
        return "\(driver.name) · \(vehicle.isEmpty ? "No vehicle" : vehicle)"
    }
}

File: pegasusX/apps/warehouse-app-android/app/src/main/java/com/pegasusx/warehouse/ui/screens/dispatch/DispatchOrderList.kt
package com.pegasusx.warehouse.ui.screens.dispatch

import androidx.compose.ui.res.stringResource

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.GridItemSpan
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.pegasusx.warehouse.data.model.DispatchPreview
import com.pegasusx.warehouse.data.model.Vehicle
import com.pegasus.design.PegasusStateKind
import com.pegasus.design.PegasusStatePane
import com.pegasusx.warehouse.ui.components.DispatchPreviewMapLibre
import com.pegasusx.warehouse.ui.components.OrderDetailOpenMode
import com.pegasusx.warehouse.ui.components.OrderOpsCard
import com.pegasusx.warehouse.ui.components.WarehouseSectionTitle
import com.pegasusx.warehouse.ui.screens.vehicles.FleetTruckDispatchCard
import com.pegasusx.warehouse.ui.theme.PegasusSpacing
import java.text.NumberFormat
import com.pegasusx.warehouse.R

private const val DISPATCH_TETRIS_BUFFER = 0.95

/**
 * Orders segment of the Dispatch screen.
 *
 * Renders the fleet truck cards, dispatch-mode selector (Smart / Manual),
 * driver picker, undispatched order list with selection checkboxes,
 * smart-suggest preview warnings, and proposed routes.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun DispatchOrderList(
    preview: DispatchPreview,
    fleetVehicles: List<Vehicle>,
    vehicleReasons: Map<String, String>,
    vehicleNotes: Map<String, String>,
    mutatingFleetVehicleId: String?,
    dispatchMode: String,
    selectedDriverId: String,
    selectedOrderIds: Set<String>,
    executing: Boolean,
    opsActingId: String?,
    driverMenuExpanded: Boolean,
    fmt: NumberFormat,
    onDispatchModeChange: (String) -> Unit,
    onDriverSelect: (String) -> Unit,
    onDriverMenuExpandChange: (Boolean) -> Unit,
    onToggleOrder: (String, Boolean) -> Unit,
    onManualDispatch: () -> Unit,
    onSmartDispatch: () -> Unit,
    onProposeDate: (String) -> Unit,
    onReject: (String) -> Unit,
    onOrderClick: (String) -> Unit,
    onVehicleClick: (String) -> Unit,
    onVehicleReasonChange: (String, String) -> Unit,
    onVehicleNoteChange: (String, String) -> Unit,
    onMarkVehicleUnavailable: (Vehicle, String, String) -> Unit,
    onRestoreVehicle: (Vehicle) -> Unit,
) {
    Column(modifier = Modifier.fillMaxSize()) {
        // ── Fleet trucks grid ──
        if (fleetVehicles.isNotEmpty()) {
            WarehouseSectionTitle(
                title = stringResource(R.string.mobile_warehouse_ui_fleet_trucks_size, fleetVehicles.size),
                modifier = Modifier.padding(horizontal = PegasusSpacing.lg, vertical = PegasusSpacing.sm),
            )
            LazyVerticalGrid(
                columns = GridCells.Adaptive(minSize = 340.dp),
                modifier = Modifier.heightIn(max = 320.dp),
                contentPadding = PaddingValues(horizontal = PegasusSpacing.lg),
                verticalArrangement = Arrangement.spacedBy(PegasusSpacing.sm),
                horizontalArrangement = Arrangement.spacedBy(PegasusSpacing.sm),
            ) {
                items(fleetVehicles, key = { it.vehicleId }) { vehicle ->
                    val selectedReason = vehicleReasons[vehicle.vehicleId]
                        ?: vehicle.unavailableReason?.takeIf { it.isNotBlank() }
                        ?: "MANUAL_HOLD"
                    val customNote = vehicleNotes[vehicle.vehicleId]
                        ?: vehicle.unavailableNote.orEmpty()
                    FleetTruckDispatchCard(
                        vehicle = vehicle,
                        selectedReason = selectedReason,
                        customNote = customNote,
                        mutating = mutatingFleetVehicleId == vehicle.vehicleId,
                        onReasonChange = { reason ->
                            onVehicleReasonChange(vehicle.vehicleId, reason)
                        },
                        onNoteChange = { note ->
                            onVehicleNoteChange(vehicle.vehicleId, note)
                        },
                        onMarkUnavailable = {
                            val note = if (selectedReason == "OTHER") customNote.trim().takeIf { it.isNotEmpty() } else null
                            onMarkVehicleUnavailable(vehicle, selectedReason, note ?: "")
                        },
                        onRestore = { onRestoreVehicle(vehicle) },
                        onOpenDetail = { onVehicleClick(vehicle.vehicleId) },
                    )
                }
            }
        }

        // ── Empty state ──
        if (preview.undispatchedOrders.isEmpty()) {
            PegasusStatePane(
                kind = PegasusStateKind.Empty,
                headline = "All orders dispatched",
                body = "No undispatched orders remain in the preview queue.",
            )
        } else {
            val selectedDriver = preview.availableDrivers.firstOrNull { it.driverId == selectedDriverId }
            val selectedVolume = preview.undispatchedOrders
                .filter { selectedOrderIds.contains(it.orderId) }
                .sumOf { it.volumeVu }
            val effectiveMax = when {
                selectedDriver?.freeVolumeVu != null && selectedDriver.freeVolumeVu > 0 ->
                    selectedDriver.freeVolumeVu * DISPATCH_TETRIS_BUFFER
                else -> (selectedDriver?.maxVolumeVu ?: 0.0) * DISPATCH_TETRIS_BUFFER
            }

            Column(modifier = Modifier.fillMaxSize()) {
                // ── Dispatch-mode selector & driver picker ──
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = PegasusSpacing.lg, vertical = PegasusSpacing.md),
                    verticalArrangement = Arrangement.spacedBy(PegasusSpacing.sm),
                ) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.spacedBy(PegasusSpacing.sm),
                    ) {
                        FilterChip(
                            selected = dispatchMode == "smart",
                            onClick = { onDispatchModeChange("smart") },
                            label = { Text("Smart fleet") },
                            modifier = Modifier.weight(1f),
                        )
                        FilterChip(
                            selected = dispatchMode == "manual",
                            onClick = { onDispatchModeChange("manual") },
                            label = { Text("Manual truck") },
                            modifier = Modifier.weight(1f),
                        )
                    }
                    if (dispatchMode == "manual") {
                    Box {
                        OutlinedButton(
                            onClick = { onDriverMenuExpandChange(true) },
                            modifier = Modifier.fillMaxWidth(),
                        ) {
                            Text(
                                selectedDriver?.let { "${it.name} · ${it.maxVolumeVu} VU max" }
                                    ?: "Select truck / driver",
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis,
                            )
                        }
                        DropdownMenu(
                            expanded = driverMenuExpanded,
                            onDismissRequest = { onDriverMenuExpandChange(false) },
                        ) {
                            preview.availableDrivers.forEach { driver ->
                                DropdownMenuItem(
                                    text = {
                                        Text(stringResource(R.string.mobile_warehouse_ui_name_truckstatus, driver.name, driver.vehicleLabel.ifBlank { driver.truckStatus }),
                                            maxLines = 1,
                                            overflow = TextOverflow.Ellipsis,
                                        )
                                    },
                                    onClick = {
                                        onDriverSelect(driver.driverId)
                                        onDriverMenuExpandChange(false)
                                    },
                                )
                            }
                        }
                    }
                    if (selectedDriver != null) {
                        Text(stringResource(R.string.mobile_warehouse_ui_loaded_format_format_2_vu_effective, "%.1f".format(selectedVolume), "%.1f".format(effectiveMax)),
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                    }
                    if (dispatchMode == "manual") {
                    Button(
                        onClick = onManualDispatch,
                        enabled = !executing && selectedDriverId.isNotBlank() && selectedOrderIds.isNotEmpty(),
                        modifier = Modifier.fillMaxWidth(),
                    ) {
                        Text(if (executing) "Dispatching…" else "Manual (${selectedOrderIds.size})")
                    }
                    } else {
                    OutlinedButton(
                        onClick = onSmartDispatch,
                        enabled = !executing && preview.undispatchedOrders.isNotEmpty() && preview.availableDrivers.isNotEmpty(),
                        modifier = Modifier.fillMaxWidth(),
                    ) {
                        Text("Smart Dispatch")
                    }
                    }
                    if (preview.fleetEffectiveCapacityVu > 0) {
                        Text(stringResource(R.string.mobile_warehouse_ui_fleet_format_vu_effective, "%.1f".format(preview.fleetEffectiveCapacityVu)),
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }

                // ── Orders grid ──
                LazyVerticalGrid(
                    columns = GridCells.Adaptive(minSize = 340.dp),
                    contentPadding = PaddingValues(horizontal = PegasusSpacing.lg, vertical = PegasusSpacing.md),
                    verticalArrangement = Arrangement.spacedBy(PegasusSpacing.md),
                    horizontalArrangement = Arrangement.spacedBy(PegasusSpacing.md),
                ) {
                items(preview.undispatchedOrders, key = { it.orderId }) { o ->
                    OrderOpsCard(
                        retailerName = o.retailerName,
                        orderId = o.orderId,
                        state = "PENDING",
                        amountLabel = fmt.format(o.totalUzs) + " ${com.pegasus.design.sessionPackCurrency()} · ${"%.1f".format(o.volumeVu)} VU",
                        showOpsMenu = true,
                        detailOpenMode = OrderDetailOpenMode.Double,
                        canDelay = true,
                        canReject = true,
                        enabled = opsActingId != o.orderId,
                        onOpenDetail = { onOrderClick(o.orderId) },
                        onDelay = { onProposeDate(o.orderId) },
                        onReject = { onReject(o.orderId) },
                        leadingContent = {
                            Checkbox(
                                checked = selectedOrderIds.contains(o.orderId),
                                onCheckedChange = { checked ->
                                    onToggleOrder(o.orderId, checked)
                                },
                            )
                        },
                    )
                }

                // ── Smart suggest preview warnings ──
                if (preview.windowConstrainedCount > 0 || preview.optimizerWarnings.isNotEmpty()) {
                    item(span = { GridItemSpan(maxLineSpan) }) {
                        Column(verticalArrangement = Arrangement.spacedBy(PegasusSpacing.xs)) {
                            WarehouseSectionTitle("Smart suggest preview")
                            if (preview.windowConstrainedCount > 0) {
                                Text(
                                    stringResource(R.string.mobile_warehouse_ui_windowconstrainedcount_order_s_constrained_by_receiving_window, preview.windowConstrainedCount),
                                    style = MaterialTheme.typography.bodySmall,
                                    color = MaterialTheme.colorScheme.tertiary,
                                )
                            }
                            preview.optimizerSource?.let { source ->
                                Text(stringResource(R.string.mobile_warehouse_ui_source_source_2, source), style = MaterialTheme.typography.bodySmall)
                            }
                            preview.optimizerWarnings.forEach { warning ->
                                Text(warning, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.tertiary)
                            }
                        }
                    }
                }

                // ── Proposed routes ──
                if (preview.proposedRoutes.isNotEmpty()) {
                    item(span = { GridItemSpan(maxLineSpan) }) {
                        Column(verticalArrangement = Arrangement.spacedBy(PegasusSpacing.sm)) {
                            WarehouseSectionTitle("Smart suggest routes (${preview.proposedRoutes.size})")
                            DispatchPreviewMapLibre(routes = preview.proposedRoutes)
                        }
                    }
                    items(preview.proposedRoutes.size) { index ->
                        val route = preview.proposedRoutes[index]
                        ElevatedCard(modifier = Modifier.fillMaxWidth()) {
                            Column(modifier = Modifier.padding(PegasusSpacing.lg)) {
                                Text(
                                    route.driverName ?: route.driverId ?: "Driver",
                                    style = MaterialTheme.typography.titleSmall,
                                )
                                Text(stringResource(R.string.mobile_warehouse_ui_size_stops_format_vu, route.stopCount ?: route.orderIds.size, "%.1f".format(route.volumeVu ?: 0.0)),
                                    style = MaterialTheme.typography.bodySmall,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                                )
                                Text(
                                    route.orderIds.joinToString(" → "),
                                    style = MaterialTheme.typography.labelSmall,
                                    maxLines = 2,
                                    overflow = TextOverflow.Ellipsis,
                                )
                            }
                        }
                    }
                }
                }
            }
        }
    }
}
